platform/openide.util.ui/src/org/openide/util/Utilities.java

```diff
@@ -1357,17 +1359,51 @@ private static Frame findMainWindow()
      */
     // PR4739
     public static Component findDialogParent() {
-        Component parent = KeyboardFocusManager.getCurrentKeyboardFocusManager().getFocusOwner();
+        return findDialogParent(null, Utilities::findMainWindow);
+    }
+
+    /**
+     * Finds an appropriate component to use for a modal dialog's parent. Similar to {@link #findDialogParent()}
+     * with the ability to specify a suggested parent component and fallback.
+     *
+     * @param suggestedParent
+     *            the component to return if non-null and valid
+     * @param fallback
+     *            a supplier for the parent if no other suitable candidate exists
+     * @return the suggested parent if there is either no active modal dialog or it is contained
+     *         within that dialog, otherwise the active modal dialog or the fallback.
+     * @since 9.30
+     */
+    public static Component findDialogParent(Component suggestedParent, Supplier<Component> fallback) {
```

Comment:
Here's a summary of my concerns/questions around `findDialogParent`, in it's current form, as a public API.

There's code scattered throughout NetBeans that does Utilities.findDialogParent(). I believe that the general consensus is that the `WindowManager`'s existing `MainWindow` is the correct default, as @matthiasblaesing first mentions in https://github.com/apache/netbeans/pull/4739#discussion_r997414384. Unfortunately, there's a dependency issue; also `WindowManager.getMainWindow` creates a main window if one doesn't exist which isn't appropriate for something in `Utilities`.

@sdedic discusses a solution in https://github.com/apache/netbeans/pull/5280#discussion_r1083271052

I'm still interestd in hearing a response to https://github.com/apache/netbeans/pull/5989#discussion_r1207270472.

Also, in `NbPresenter` there's `findFocusedWindow` which has a hardcoded string `NbMainWindow`

The only use of the proposed `findDialogParent` is `findDialogParent(null, WindowManager.getDefault()::getMainWindow)` in `DialogDisplayImpl`; this feels like a bandaid/workaround for the inability of Utilities to find the main window. As a bandiad, should it become a public API? There is a workaound for `DialogDisplayImpl`, using this PR's previous version, something like
```
    if((parentComponent = Utilities.findDialogParent(null)) == null)
        parentComponent = WindowManager.getDefault()::getMainWindow);
```

What's the use case for providing a fallback? Other than working around a know deficiency with getting the main window?

There is a general problem with finding the default main window; if this problem were addressed, would it affect this new public API? And/or the need for it?   Should the problem with finding the main window be fixed (or at least have a target solution) before there's a public API around this issue. Maybe they're orthoginal issues, it's not clear to me.

In general, this is a beautiful PR; but, IMHO, this new findDialolgParent public API seems mis-directed.

Reply:
Forgot to mention @neilcsmith-net 's https://github.com/apache/netbeans/pull/5280#discussion_r1073899226 about finding the "real" main window and not using a hardcoded string.

Reply:
> this new `findDialolgParent` public API seems mis-directed

My primary goal in adding this was to allow for a dialog to be positioned relative to a component within the currently active window (assuming there isn't another modal window open) rather than to the whole window. This is similar to the way the parent component is used in `JOptionPane` (see `JOptionPane.initDialog(JDialog, int, Component)`).

If it is more palatable, I can remove the `Supplier` argument and just have it default to the main window which would connect it more closely to the platform rather than being a totally generic method.

Reply:
I'm confused. This statement seems innacurate.

> adding this was to allow for a dialog to be positioned relative to a component within the
> currently active window (assuming there isn't another modal window open) rather than to the
> whole window.

If there's a "currently active window" (or a focused component), then the fallback is not used.

And I couldn't find `JOptionPane.initDialog()`.

AFAICT, the fallback is only used when there is _no_ focus owner and _no_ active window, and no
dialog. I think this is a rare and unusual case; I'm not sure I know how to create it.

Maybe there's some miscommunication. The issues I raised, stated in the first line, are about the
`findDialogParent` **in its current form**, which means the fallback.

I want to understand the use-case for it; in particular, is it only to work around a NB bug?